burotel/indico_burotel/tasks.py

```diff
@@ -0,0 +1,188 @@
+# the LICENSE file for more details.
+
+
+import requests
+from celery.exceptions import Retry
+from celery.schedules import crontab
+from flask_pluginengine import current_plugin
+from requests.exceptions import RequestException, Timeout
+from sqlalchemy.types import Interval
+
+from indico.core.auth import multipass
+from indico.core.celery import celery
+from indico.core.db import db
+from indico.modules.rb.models.reservations import Reservation, ReservationEditLog, ReservationState
+from indico.modules.rb.models.room_attributes import RoomAttribute, RoomAttributeAssociation
+from indico.modules.rb.models.rooms import Room
+from indico.modules.users.models.users import User
+from indico.util.date_time import now_utc
+
+from indico_burotel.notifications import notify_about_to_cancel, notify_automatic_cancellation
+
+
+def _get_cern_ident(plugin, user):
+    cern_ident_provider = plugin.settings.get('cern_identity_provider')
+    cern_ident = next(
+        (ident for ident in user.identities if ident.provider == cern_ident_provider),
+        None
+    )
+    if not cern_ident or not cern_ident.multipass_data or not cern_ident.multipass_data.get('cern_person_id'):
+        return None
+    else:
+        return cern_ident
+
+
+def _find_person_id(plugin, user):
+    """Get the CERN Person ID of a user."""
+    cern_ident_provider = plugin.settings.get('cern_identity_provider')
+    cern_ident = _get_cern_ident(plugin, user)
+
+    if not cern_ident:
+        # try to fetch the identity from the source before giving up
+        results = multipass.search_identities(providers={cern_ident_provider}, exact=True, email=user.all_emails)
+        if not (cern_ident := next(results, None)):
+            # that didn't work out, raise an error
+            raise Exception("Can't find %s identity for %s", cern_ident_provider, user)
+
+    return cern_ident.multipass_data.get('cern_person_id')
+
+
+def _build_query(delta):
+    attr = RoomAttribute.query.filter(RoomAttribute.name == 'confirmation-by-secretariat').one()
+    reservation = db.aliased(Reservation)
+
+    # get all days between two dates
+    subq = db.session.query(db.func.generate_series(
+        reservation.start_dt,
+        now_utc(),
+        db.func.cast('1 day', Interval)
+    ).label('d')).subquery()
+
+    # count all days except for weekends (working days)
+    query = db.session.query(
+        db.func.count(subq.c.d).label('d')
+    ).filter(db.func.extract('dow', subq.c.d).notin_({0, 6})).subquery()
+
+    return reservation.query.filter(
+        reservation.state == ReservationState.pending,
+        # booking ends in the future
+        reservation.end_dt > now_utc(),
+        # booking has started more than `delta` working days ago
+        query.c.d > delta,
+        Room.attributes.any(
+            db.and_(
+                RoomAttributeAssociation.attribute == attr,
+                RoomAttributeAssociation.value.isnot(None)
```

Comment:
should we check for something explicit like `yes`? I can see people putting `no` or `false` in the attributes ;x

(in the future it'd be nice to set a type for attributes and then show different UI for bool attributes, and storing actual bools in the json value)

Reply:
Good idea.